Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/ReflectometryILLPreprocess.py

```diff
@@ -397,67 +350,111 @@ def _foregroundWidths(self):
+                Filename=inputFiles,
+                LoaderName='LoadILLReflectometry',
+                LoaderVersion=-1,
+                LoaderOptions=loadOption,
+                OutputWorkspace=mergedWSName,
+                EnableLogging=self._subalgLogging
+            )
+        else:
+            ws = self.getProperty(Prop.INPUT_WS).value
+            self._cleanup.protect(ws)
+        return ws
+
+    def _peakFitting(self, ws):
+        """Add peak and foreground information to the sample logs."""
+        # Convert to wavelength
+        l1 = ws.spectrumInfo().l1()
+        hists = ws.getNumberHistograms()
+        mindex = int(hists / 2)
+        l2 = ws.spectrumInfo().l2(mindex)
+        theta = ws.spectrumInfo().twoTheta(mindex) / 2.
+        linePosition = 0.0
+        if not self.getProperty(Prop.LINE_POSITION_INPUT).isDefault:
+            linePosition = self.getProperty(Prop.LINE_POSITION_INPUT).value
         else:
-            if not self.getProperty(Prop.BEAM_CENTRE).isDefault:
-                peakPos = self.getProperty(Prop.BEAM_CENTRE).value
-                beamPosWS = self._createFakePeakPositionTable(peakPos)
-            else:
-                # Beam position will be fitted later.
-                beamPosWS = None
-        return ws, beamPosWS
+            # Fit peak position
+            peakWSName = self._names.withSuffix('peak')
+            xmin = self.getProperty(Prop.XMIN).value
+            if not xmin == Property.EMPTY_DBL:
+                xmin = self.getProperty(Prop.XMIN).value
+                xmin = common.inTOF(xmin, l1, l2, theta)
+            xmax = self.getProperty(Prop.XMAX).value
+            if not self.getProperty(Prop.XMAX).isDefault:
+                xmax = self.getProperty(Prop.XMAX).value
+                xmax = common.inTOF(xmax, l1, l2, theta)
+            peak = FindReflectometryLines(
+                InputWorkspace=ws,
+                OutputWorkspace=peakWSName,
+                RangeLower=xmin,
+                RangeUpper=xmax,
+                StartWorkspaceIndex=self.getProperty(Prop.START_WS_INDEX).value,
+                EndWorkspaceIndex=self.getProperty(Prop.END_WS_INDEX).value
+            )
+            linePosition = peak.LineCentre
+            self._cleanup.cleanup(peak.OutputWorkspace)
+        return ws, linePosition
+
+    def _addSampleLogInfo(self, ws, linePosition):
+        """Add foreground indices and linePosition to the sample logs, names start with reduction."""
+        # Add the fractional workspace index of the beam position to the sample logs of ws.
+        run = ws.run()
+        run.addProperty(common.SampleLogs.LINE_POSITION, float(linePosition), True)
+        # Add foreground start and end workspace indices to the sample logs of ws.
+        hws = self._foregroundWidths()
+        beamPosIndex = int(numpy.rint(linePosition))
+        if beamPosIndex > 255:
+            beamPosIndex = 255
+            self.self.log().warning('Is it a monitor spectrum?')
```

Comment:
Perhaps the warning message could be improved. On a second thought, is this condition even possible? The input properties should be validated between 0 and 255 and fitting should return something within that interval as well, or am I missing something?

Reply:
The user given line position must be between 0.0 and 255.0 by validation. The start and end workspace indices can be validated to be greater than 0.0 and smaller than 255.0 for start and end workspace index, respectively. In this case, the above check can be deleted. Seems to be a better way.